Add `xargo-check` command

Aaron1011 opened the pull request:
This allows configuring Xargo to run `cargo check`
instead of `cargo build`

Needed to support https://github.com/rust-lang/miri/pull/1048

Aaron1011 commented:
cc @RalfJung

RalfJung commented on README.md:
```diff
@@ -309,6 +309,21 @@ git = "https://github.com/japaric/steed"
 
+## Cargo mode
+
+Xaro supports configuring the mode in which `cargo` is invoked.
```
Specifically this affects only the sysroot cargo, right? Typical use of xargo involves also having an actual project that is then built once libstd is done.

RalfJung commented:
Thanks a lot! I wanted some way to make a check-only libstd build for some time but did not get around to actually work on this.

However, I am not sure what the best UI is for this. Originally I envisioned that `xargo check` would also do a check-only libstd build, but that is likely not going to work very well. So a toml entry seems fine to me. @jethrogb what do you think (or do you not care)?

In terms of this PR, please add a test to make sure that this actually works.

RalfJung commented on README.md:
````diff
@@ -309,6 +309,21 @@ git = "https://github.com/japaric/steed"
+This is set via the `cargo_mode` entry in `Xargo.toml`:
+
+``` toml
+cargo_mode = "check"
````
"mode" is an odd choice of word, I think cargo calls this a "command". Also since this only affects the sysroot, maybe something like `sysroot_cargo_command` or so would be better?

jethrogb commented:
> Originally I envisioned that `xargo check` would also do a check-only libstd build, but that is likely not going to work very well.

This should work, right? As long as all the metadata is there...

That would be my preferred solution.

If we want to go the way this PR is written now, since this doesn't produce a usable sysroot, I think it's ok if it's invoked differently from regular `xargo build`. Like `xargo --check-std`. I don't think it makes a lot of sense to put it as an option in `Xargo.toml`. Do we parse the cmdline right now?

Aaron1011 commented:
> Do we parse the cmdline right now?

Currently, I think all arguments are forwarded to Cargo. Personally, I would prefer it if we kept it that way.

RalfJung commented:
> This should work, right? As long as all the metadata is there...

I don't think so; even doing `cargo check` will *build* the build scripts and proc macros, and that needs a libstd with code, not just its metadata.

RalfJung commented:
@jethrogb
> since this doesn't produce a usable sysroot, I think it's ok if it's invoked differently from regular xargo build. Like xargo --check-std. I don't think it makes a lot of sense to put it as an option in Xargo.toml. 

As an alternative option, what about using an env var to control this?

jethrogb commented:
> I don't think so; even doing cargo check will build the build scripts and proc macros, and that needs a libstd with code, not just its metadata.

No it only needs that for the host, not the target

> As an alternative option, what about using an env var to control this?

That also seems somewhat weird to me. What about a second binary `xargo-check`?

RalfJung commented:
> No it only needs that for the host, not the target

For our use-case (Miri), those are the same.

Aaron1011 commented:
I'd be fine with a second binary. To be clear, this would be invoked as `xargo-check` to avoid conflicting with `xargo check`, right?

Aaron1011 commented:
@RalfJung @jethrogb: I've created a new `xargo-check` command, and added a test for it. The old `xargo.rs` was moved into `lib.rs`. The `bin/xargo.rs` and `bin/xargo-check.rs` files simply pass in the proper mode to the main entrypoint.

RalfJung commented on README.md:
````diff
@@ -309,6 +309,18 @@ git = "https://github.com/japaric/steed"
 stage = 2
 ```
 
+## Cargo subcommand
````
I don't entire understand the title here. What about something like "Check-only libstd build"?

Aaron1011 replied:
This was left over from the original incarnation of this PR. I'll change it to something more suitable.

RalfJung commented on src/lib.rs:
```diff
@@ -187,7 +189,8 @@ fn run() -> Result<ExitStatus> {
                 verbose,
-                args.message_format()
+                args.message_format(),
+                cargo_mode
```
```suggestion
                cargo_mode,
```

RalfJung commented on src/sysroot.rs:
```diff
@@ -38,7 +38,8 @@ fn build(
     verbose: bool,
-    message_format: Option<&str>
+    message_format: Option<&str>,
+    cargo_mode: CargoMode
```
```suggestion
    cargo_mode: CargoMode,
```

RalfJung commented on src/sysroot.rs:
```diff
@@ -241,7 +245,8 @@ pub fn update(
     verbose: bool,
-    message_format: Option<&str>
+    message_format: Option<&str>,
+    cargo_mode: CargoMode
```
```suggestion
    cargo_mode: CargoMode,
```

RalfJung commented on src/sysroot.rs:
```diff
@@ -268,6 +273,7 @@ pub fn update(
             hash,
             verbose,
             message_format,
+            cargo_mode
```
```suggestion
            cargo_mode,
```

RalfJung commented on src/sysroot.rs:
```diff
@@ -320,6 +326,14 @@ pub struct Stage {
+#[derive(Copy, Clone, Debug)]
+pub enum CargoMode {
+    Build,
+    Check
```
```suggestion
    Check,
```

RalfJung commented on tests/smoke.rs:
```diff
@@ -336,6 +345,15 @@ impl HProject {
         write(&self.td.path().join("Xargo.toml"), false, toml)
     }
+
+    fn check(&self) -> Result<()> {
```
This sounds a bit like it would do `xargo check` as opposed to `xargo-check`.

RalfJung commented:
> I've created a new xargo-check command, and added a test for it. The old xargo.rs was moved into lib.rs. The bin/xargo.rs and bin/xargo-check.rs files simply pass in the proper mode to the main entrypoint.

The approach looks good! I left some nits.

One higher-level point: is there any point to actually run the per-project `cargo` after the libstd check build is done? That build will anyway probably not work as expected. So it might make more sense to make `xargo-check` *only* build libstd, and nothing else?

Aaron1011 commented:
> So it might make more sense to make xargo-check only build libstd, and nothing else?

@RalfJung: You can already do this by running `xargo-check` with no additional arguments.

RalfJung commented:
> You can already do this by running `xargo-check` with no additional arguments.

Oh wow, I had no idea xargo supports this.
Could you adjust the new test to do that instead?

Aaron1011 commented:
> EDIT: Hm, but looks like that invokes cargo's help screen? (At least it does for the main xargo binary.)

You need to be in a project directly (with a `Cargo.toml`) for this to work.

RalfJung commented:
Yes that's what I did.

Aaron1011 commented:
That's strange - it works for me locally. I get the help screen when outside a project directly, but a proper built when inside the directory.

I've created tests for both `xargo-check` and `xargo-check check`, which both pass locally. Let's see if they pass on CI.

RalfJung commented:
I get a build followed by the help screen. And given that there is no logic in xargo for *not* invoking `cargo` when there is no command, that's what I would expect.

Can you make `xargo-check` not call `cargo` for the project when there is no subcommand?

Aaron1011 commented:
@RalfJung: I think it's better to have `xargo` and `xargo-check` be identical, except for how they build the sysroot. Since `xargo` (with no additional arguments) displays the help screen after building the sysrooot (which I think is useful), I think `xargo-check` should as well.

RalfJung commented:
Given that `xargo-check` produces a basically broken sysroot, I am not convinced there is any reason to even try and run cargo afterwards. The two are already inconsistent in that regard.

Aaron1011 commented:
@RalfJung: We'll want to run to run `cargo` afterwards when we build the sysroot in Miri, though. I don't see why we would want to reduce the flexibility of `xargo-check`.

RalfJung commented:
> We'll want to run cargo afterwards when we build the sysroot in Miri, though.

No, definitely not. Why would we?

Aaron1011 commented:
@RalfJung: As discussed in https://github.com/rust-lang/miri/issues/1057, we'll need to switch from `cargo rustc` to `cargo check` in Miri, since `cargo rustc` will try to do a normal build. We can therefore just make one call (`xargo-check check`), which will do both a check build of the sysroot and the dependencies.

> Given that xargo-check produces a basically broken sysroot, I am not convinced there is any reason to even try and run cargo afterwards. 

`xargo-check` is such a specialized command - I'm not even sure that anyone outside of Miri will find a use for it. I see no reason to artificially limit `xargo-check` compared to `xargo` (e.g. adding extra logic to disable the behavior we currently get for free). Anyone invoking `xargo-check` already needs to know what they're doing - I don't think preventing subcommands from running makes it any less of a footgun for people who mean to run `xargo check`.

RalfJung commented:
> We can therefore just make one call (xargo-check check), which will do both a check build of the sysroot and the dependencies.

I don't think we want that. (a) we don't want the `Xargo.toml` to be in the project directory. (b) We want to skip running xargo if `MIRI_SYSROOT` is set, and I'd rather not have two different code paths here.

> xargo-check is such a specialized command - I'm not even sure that anyone outside of Miri will find a use for it.

That's an argument *for* not displaying a help screen when no subcommand is given -- that is the best way to cater Miri's use case.

jethrogb commented:
I expect `xargo-check` to work in a directory that contains only a `Xargo.toml`. Is that the case?

jethrogb commented:
> xargo-check is such a specialized command - I'm not even sure that anyone outside of Miri will find a use for it. 

`xargo` has been known to be used for `std` development for any platform, and I think this tool fits that use case nicely.

Aaron1011 commented:
> I expect xargo-check to work in a directory that contains only a Xargo.toml. Is that the case?

Yes, it is.

> xargo has been known to be used for std development for any platform, and I think this tool fits that use case nicely.

I don't know if a check-build of `libstd` is even usable outside of Miri - we only require that the MIR exist, and don't test whether it works from a normal Cargo compilation. I suppose someone could be working on the standard library itself - but in that case, they'd want to use `x.py` in their local checkout of rustc.

Aaron1011 commented:
@RalfJung 

> We want to skip running xargo if MIRI_SYSROOT is set, and I'd rather not have two different code paths here.

I didn't realize that `MIRI_SYSROOT` was a thing - that's a good point.

> That's an argument for not displaying a help screen when no subcommand is given -- that is the best way to cater Miri's use case.

I still think it's better to keep `xargo` and `xargo-check` as consistent as possible, even if Miri only ever runs a plain `xargo-check`.

From my perspective, the changes you're proposing to `xargo-check` (omitting the help message, and not running cargo) don't help with the current Miri usecase, and only serve to potentially hinder other projects using `xargo`.

Given that we're adding a new command, I think the risk of people accidentally getting a broken setup (e.g. running `xargo-check build` instead of `xargo build`) is essentially zero. They would have to either discover `xargo-check` on their own, or read the README and ignore the warning about `xargo-check`.

jethrogb commented:
> I suppose someone could be working on the standard library itself - but in that case, they'd want to use x.py in their local checkout of rustc.

When using x.py you need to build all of stage0, so using xargo is much faster.

Aaron1011 commented:
> When using x.py you need to build all of stage0, so using xargo is much faster.

That's true - maybe `xargo-check` could actually be useful for that usecase. I think that's a further reason not to artificially limit `xargo-check` - if someone wants to experiment with it, they should be able to.

RalfJung commented:
> >   I expect xargo-check to work in a directory that contains only a Xargo.toml. Is that the case?

> Yes, it is.

Really? It'll show the help message, indicating incorrect usage. And `xargo-check check` will, I assume, not work without a `Cargo.toml`.

So, as far as I am concerned, `xargo-check` does not work (properly) in a directory with only a `Xargo.toml` and no `Cargo.toml`.

Either way, could you add a test for that case?

> Given that we're adding a new command, I think the risk of people accidentally getting a broken setup (e.g. running xargo-check build instead of xargo build) is essentially zero.

The fact that it is a new command also means that total consistency for corner cases is less relevant.

>  I think that's a further reason not to artificially limit xargo-check

Not sure what you mean by "limit" here. Working on libstd is just yet another use case that doesn't want a `Cargo.toml`, meaning that should be properly supported.

But anyway, `./x.py build --stage 0 src/libstd` is very fast, so I don't think this is a useful use-case for xargo anyway.

RalfJung commented:
I just tried the following:

* Have a dir with just the following file in it:
```
[dependencies.std]
default_features = false
# We need the `panic_unwind` feature because we use the `unwind` panic strategy.
# Using `abort` works for libstd, but then libtest will not compile.
features = ["panic_unwind"]

[dependencies.test]
```
* Run `xargo-check` in that dir.

The result is that xargo prints a help screen and does nothing else. In particular, nothing gets built. So, to me it doesn't look like xargo-check actually works with just a `Xargo.tom`?

When I add a dummy `Cargo.toml`+`lib.rs`, the build happens as expected (followed by a help screen).

Aaron1011 commented:
> So, as far as I am concerned, xargo-check does not work (properly) in a directory with only a Xargo.toml and no Cargo.toml.

> Either way, could you add a test for that case?

Do you want `xargo-check` to support running in a directory with only an `Xargo.toml`, or do you just want a test verifying the current behavior?

> Not sure what you mean by "limit" here. Working on libstd is just yet another use case that doesn't want a Cargo.toml, meaning that should be properly supported.

@RalfJung: By "limit", I mean introducing logic that restricts running subcommands via `xargo-check`, when it would otherwise work.

jethrogb commented:
> Do you want xargo-check to support running in a directory with only an Xargo.toml

I asked for this and you said it was supported.

Aaron1011 commented:
@jethrogb: I was mistaken - I didn't remove the `Cargo.toml` before I tested.

RalfJung commented:
> Do you want xargo-check to support running in a directory with only an Xargo.toml

Yes.

> By "limit", I mean introducing logic that restricts running subcommands via xargo-check, when it would otherwise work.

Okay. Supporting subcommands when a `Cargo.toml` is present seems reasonable. But IMO there should also be a "proper" way to invoke xargo-check without a `Cargo.toml`.

Aaron1011 commented:
> But IMO there should also be a "proper" way to invoke xargo-check without a Cargo.toml.

Sounds good to me

Aaron1011 commented:
@RalfJung: I've modified `xargo-check` to support running in a directory with only an `Xargo.toml`. I added an additional test to verify that this works correctly.

RalfJung commented on tests/smoke.rs:
```diff
@@ -887,3 +908,44 @@ tag = "1.0.25"
+fn cargo_check_plain() {
+    fn run() -> Result<()> {
+        let project = HProject::new(false)?;
+        project.xargo_check_subcommand(None)?;
```
`cargo_check_check_no_ctoml` already covers this, I'd say.

RalfJung commented on tests/smoke.rs:
```diff
@@ -887,3 +908,44 @@ tag = "1.0.25"
+fn cargo_check_check_no_ctoml() {
+    fn run() -> Result<()> {
+        let project = HProject::new(false)?;
+        project.xargo_toml("")?;
```
So I guess `HProject` does not per default create a `Xargo.toml`? If so, please add a comment here like "make sure Xargo.toml exists" or so.

RalfJung commented:
I checked this locally and `xargo-check` now looks like this:
```
$ rm ~/.xargo/ -rf && xargo-check
   Compiling compiler_builtins v0.1.22
   Compiling cc v1.0.47
    Checking core v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore)
   Compiling build_helper v0.1.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/build_helper)
   Compiling libc v0.2.64
   Compiling autocfg v0.1.6
   Compiling std v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd)
   Compiling cmake v0.1.38
   Compiling hashbrown v0.6.2
   Compiling unwind v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libunwind)
   Compiling backtrace-sys v0.1.32
   Compiling rustc_lsan v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/librustc_lsan)
   Compiling rustc_asan v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/librustc_asan)
   Compiling rustc_tsan v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/librustc_tsan)
   Compiling rustc_msan v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/librustc_msan)
    Checking rustc-std-workspace-core v1.99.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/tools/rustc-std-workspace-core)
    Checking alloc v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/liballoc)
    Checking cfg-if v0.1.8
    Checking rustc-demangle v0.1.16
    Checking panic_abort v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libpanic_abort)
    Checking backtrace v0.3.40
    Checking rustc-std-workspace-alloc v1.99.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/tools/rustc-std-workspace-alloc)
    Checking panic_unwind v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libpanic_unwind)
    Finished release [optimized] target(s) in 17.09s
    Checking rustc-std-workspace-std v1.99.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/tools/rustc-std-workspace-std)
    Checking term v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libterm)
    Checking proc_macro v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libproc_macro)
    Checking unicode-width v0.1.6
    Checking getopts v0.2.21
    Checking test v0.0.0 (/home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest)
    Finished release [optimized] target(s) in 2.14s
Rust's package manager

USAGE:
    cargo [OPTIONS] [SUBCOMMAND]

OPTIONS:
    -V, --version           Print version info and exit
        --list              List installed commands
        --explain <CODE>    Run `rustc --explain CODE`
    -v, --verbose           Use verbose output (-vv very verbose/build.rs output)
    -q, --quiet             No output printed to stdout
        --color <WHEN>      Coloring: auto, always, never
        --frozen            Require Cargo.lock and cache are up to date
        --locked            Require Cargo.lock is up to date
        --offline           Run without accessing the network
    -Z <FLAG>...            Unstable (nightly-only) flags to Cargo, see 'cargo -Z help' for details
    -h, --help              Prints help information

Some common cargo commands are (see all commands with --list):
    build       Compile the current package
    check       Analyze the current package and report errors, but don't build object files
    clean       Remove the target directory
    doc         Build this package's and its dependencies' documentation
    new         Create a new cargo package
    init        Create a new cargo package in an existing directory
    run         Run a binary or example of the local package
    test        Run the tests
    bench       Run the benchmarks
    update      Update dependencies listed in Cargo.lock
    search      Search registry for crates
    publish     Package and upload this package to the registry
    install     Install a Rust binary. Default location is $HOME/.cargo/bin
    uninstall   Uninstall a Rust binary

See 'cargo help <command>' for more information on a specific command.
```
Given taht building *just* libstd is now an intended usecase, I don't think it is a good idea to show the cargo help screen there. `xargo` should just not invoke local `cargo` at all when there is no `Cargo.toml`.

Aaron1011 commented:
@RalfJung: Updated

RalfJung commented on src/sysroot.rs:
```diff
@@ -344,6 +364,14 @@ pub struct Stage {
+/// Which mode to invoke `cargo` in when building the sysroot
+/// Can be either `cargo build` or `cargo check`
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum CargoMode {
```
I think `XargoMode` or `SysrootMode` might be a better name for this... because this is no the mode used for the project build, right, just the sysroot?

RalfJung commented:
@Aaron1011 thanks, that looks good! There are still two open comments from before though.

Aaron1011 commented:
@RalfJung: I've addressed your comments

RalfJung commented:
Thanks!

bors r+

RalfJung commented:
Dang, CI fails because `default-run` was not stable yet in nightly-2018-12-01 (which we test for backwards compatibility).

CI itself doesn't need that feature, right? So we could `sed`-patch `Cargo.toml` to make it compatible with the old nightly on CI? That's quite hacky, but I cannot think of anything more reasonable right now.

Aaron1011 commented:
@RalfJung: I've added the hack

RalfJung commented:
bors r+

RalfJung commented:
Looks like that sed script doesn't work on macOS :(

jethrogb commented:
https://stackoverflow.com/questions/4247068/sed-command-with-i-option-failing-on-mac-but-works-on-linux

Aaron1011 commented:
@RalfJung: I've now just using a plain `sed` invocation, along with `mv`

jethrogb commented:
bors r=RalfJung
